OpenRA.Game/Network/UPnP.cs

```diff
@@ -10,127 +10,77 @@
-				NatUtility.StartDiscovery();
-				Log.Write("server", "NAT discovery started.");
-			}
-			catch (Exception e)
-			{
-				Log.Write("server", "Can't discover UPnP-enabled device: {0}", e);
-				Game.Settings.Server.NatDeviceAvailable = false;
-				Game.Settings.Server.AllowPortForward = false;
-			}
-		}
+		public static IPAddress ExternalIP;
 
-		public static void StoppingNatDiscovery()
+		public static async Task DiscoverNatDevices(int timeout)
 		{
-			Log.Write("server", "Stopping NAT discovery.");
-			NatUtility.StopDiscovery();
-
-			if (natDevice == null || natDevice.GetType() != typeof(Mono.Nat.Upnp.UpnpNatDevice))
-			{
-				Log.Write("server",
-					"No NAT devices with UPnP enabled found within {0} ms deadline. Disabling automatic port forwarding.".F(Game.Settings.Server.NatDiscoveryTimeout));
-				Game.Settings.Server.NatDeviceAvailable = false;
-				Game.Settings.Server.AllowPortForward = false;
-			}
-		}
-
-		public static void DeviceFound(object sender, DeviceEventArgs args)
-		{
-			Log.Write("server", "NAT device discovered.");
-
-			Game.Settings.Server.NatDeviceAvailable = true;
-			Game.Settings.Server.AllowPortForward = true;
-
-			try
-			{
-				natDevice = args.Device;
-				Log.Write("server", "Type: {0}", natDevice.GetType());
-				Log.Write("server", "Your external IP is: {0}", natDevice.GetExternalIP());
-
-				foreach (var mp in natDevice.GetAllMappings())
-					Log.Write("server", "Existing port mapping: protocol={0}, public={1}, private={2}",
-						mp.Protocol, mp.PublicPort, mp.PrivatePort);
-			}
-			catch (Exception e)
-			{
-				Log.Write("server", "Can't fetch information from NAT device: {0}", e);
+			NatDiscoverer.TraceSource.Switch.Level = SourceLevels.Verbose;
+			Log.AddChannel("nat", "nat.log");
+			var logChannel = Log.Channel("nat");
+			NatDiscoverer.TraceSource.Listeners.Add(new TextWriterTraceListener(logChannel.Writer));
 
-				Game.Settings.Server.NatDeviceAvailable = false;
-				Game.Settings.Server.AllowPortForward = false;
-			}
-		}
-
-		public static void ForwardPort(int lifetime)
-		{
-			try
-			{
-				var mapping = new Mapping(Protocol.Tcp, Game.Settings.Server.ExternalPort, Game.Settings.Server.ListenPort, lifetime);
-				natDevice.CreatePortMap(mapping);
-				Log.Write("server", "Create port mapping: protocol = {0}, public = {1}, private = {2}, lifetime = {3} s",
-					mapping.Protocol, mapping.PublicPort, mapping.PrivatePort, mapping.Lifetime);
-			}
-			catch (MappingException e)
+			var natDiscoverer = new NatDiscoverer();
+			var token = new CancellationTokenSource(timeout);
+			natDevices = await natDiscoverer.DiscoverDevicesAsync(PortMapper.Upnp, token);
+			foreach (var natDevice in natDevices)
 			{
-				if (e.ErrorCode == 725 && lifetime != 0)
+				try
 				{
-					Log.Write("server", "NAT device answered with OnlyPermanentLeasesSupported. Retrying...");
-					ForwardPort(0);
+					ExternalIP = natDevice.GetExternalIPAsync().Result;
```

Comment:
`await natDevice.GetExternalIPAsync();` would be better.

Also, should we break after we set the ExternalIP? It doesn't seem like we need to query every device if one of them responds.